flang/lib/Lower/OpenMP/DataSharingProcessor.cpp

```diff
@@ -308,7 +308,10 @@ void DataSharingProcessor::insertLastPrivateCompare(mlir::Operation *op) {
     firOpBuilder.setInsertionPointToStart(&ifOp.getThenRegion().front());
     for (auto [v, loopIV] : llvm::zip_equal(vs, loopIVs)) {
       assert(loopIV && "loopIV was not set");
-      firOpBuilder.createStoreWithConvert(loc, v, loopIV);
+      hlfir::Entity loopIVEntity{loopIV};
+      loopIVEntity =
+          hlfir::derefPointersAndAllocatables(loc, firOpBuilder, loopIVEntity);
+      firOpBuilder.create<hlfir::AssignOp>(loc, v, loopIVEntity);
```

Comment:
The lowering code for hlfir:Assign to fir::store seems to insert the convert, hence skipping that here.